### PR TITLE
fs: Do not report error when errors were fixed by e2fsck

### PR DESCRIPTION
**Pull request by vojtechtrefny**
Fixes: #587

### CODE REVIEW

**Comment by vojtechtrefny** [comment]
Question: What should we do about "File system errors corrected, system should be rebooted"? I don't think we should completely ignore this but showing error also doesn't sound like the best thing to do. Maybe a special error state that UDisks (and GNOME Disks) can catch and show just an "info" message to the user?

**Comment by tbzatek** [comment]
> Question: What should we do about "File system errors corrected, system should be rebooted"? I don't think we should completely ignore this but showing error also doesn't sound like the best thing to do. Maybe a special error state that UDisks (and GNOME Disks) can catch and show just an "info" message to the user? @tbzatek @vpodzime

Hmm, I think this is typically designed for init systems. Right now we don't have a way to display a hint, we can log it however (through `g_syslog()` or so) and let some potential (currently nonexisting) storage notification daemon to pick it up. I would not fail the operation as it breaks the semantic of the `gboolean` return value and not every consumer is watching for error codes.

Still, in case `e2fsck` returns such code, would kernel allow mounting that fs afterwards? This should be probably handled later at the mount process, `e2fsck` is only changing data on a block device, unless it can operate on a mounted filesystem.

File: src/plugins/fs/ext.c [hunk 1]
```diff
@@ -410,15 +410,22 @@ static gboolean ext_repair (const gchar *device, gboolean unsafe, const BDExtraA
...
-        return bd_utils_exec_and_report_progress (args_progress, extra, extract_e2fsck_progress, &status, error);
+        ret = bd_utils_exec_and_report_progress (args_progress, extra, extract_e2fsck_progress, &status, error);
     } else {
-        return bd_utils_exec_and_report_status_error (args, extra, &status, error);
+        ret = bd_utils_exec_and_report_status_error (args, extra, &status, error);
     }
+
+    if (!ret && (status == 1)) {
+        /* no error should be reported for exit code 1 - File system errors corrected */
+        g_clear_error (error);
```
Comment (tbzatek) [inline src/plugins/fs/ext.c:426]: Then `ret` should be set accordingly.

Comment (vojtechtrefny) [reply src/plugins/fs/ext.c:426]: Good catch, fixed.

**Review by tbzatek — APPROVED** [review]

**Comment by vojtechtrefny** [comment]
> Right now we don't have a way to display a hint, we can log it however (through `g_syslog()` or so) and let some potential (currently nonexisting) storage notification daemon to pick it up.

I was thinking about returning special error from libblockdev so UDisks can return a special error for GNOME Disks to catch.

> Still, in case `e2fsck` returns such code, would kernel allow mounting that fs afterwards? This should be probably handled later at the mount process, `e2fsck` is only changing data on a block device, unless it can operate on a mounted filesystem.

`e2fsck` can run on mounted devices but it's not recommend. Not sure how `mount` behaves when this happens.

**Review by vpodzime — APPROVED** [review]
Looks good to me. I think the reboot hint is outside of the scope of the code here. Like @tbzatek suggested, it could be logged so that some future storage monitoring system grabs it and acts on it, but that's all. And I wouldn't bother with that now. If such storage monitoring system ever exists, it should probably be *e2fsck* itself communicating with it somehow (through the log messages, for example).

**Comment by tbzatek** [comment]
> I was thinking about returning special error from libblockdev so UDisks can return a special error for GNOME Disks to catch.

Take it from a pragmatic point of view: is the filesystem fixed? It is.

Returning an error (and `False` return code) would create unnecessary confusion to legacy consumers. If we set `error` and return `True`, it would create new memleaks. Besides that would qualify as an ABI break, though it's not that important here I guess.

**Comment by vojtechtrefny** [comment]
> > I was thinking about returning special error from libblockdev so UDisks can return a special error for GNOME Disks to catch.
> 
> Take it from a pragmatic point of view: is the filesystem fixed? It is.
> 
> Returning an error (and `False` return code) would create unnecessary confusion to legacy consumers. If we set `error` and return `True`, it would create new memleaks. Besides that would qualify as an ABI break, though it's not that important here I guess.

So I should change the code to return `True` when reboot is needed? Because right now it returns `False` and sets the error. (The change fixed only `1 - File system errors corrected` and not `2    - File system errors corrected, system should be rebooted`.)

**Comment by tbzatek** [comment]
> So I should change the code to return `True` when reboot is needed? Because right now it returns `False` and sets the error. (The change fixed only `1 - File system errors corrected` and not `2 - File system errors corrected, system should be rebooted`.)

Yes, I think so. Both the `1` and `2` return codes should be treated equally as _"filesystem has been fixed"_ and return `True` IMO.

**Comment by vpodzime** [comment]
> Yes, I think so. Both the `1` and `2` return codes should be treated equally as _"filesystem has been fixed"_ and return `True` IMO.

Yup, agreed.

**Comment by vojtechtrefny** [comment]
> > So I should change the code to return `True` when reboot is needed? Because right now it returns `False` and sets the error. (The change fixed only `1 - File system errors corrected` and not `2 - File system errors corrected, system should be rebooted`.)
> 
> Yes, I think so. Both the `1` and `2` return codes should be treated equally as _"filesystem has been fixed"_ and return `True` IMO.

Ok, I've updated the PR.

**Review by vpodzime — APPROVED** [review]
ACK

**Pull request merged**